kimchi/src/verifier_index.rs

```diff
@@ -325,4 +329,25 @@ where
         self.serialize(&mut rmp_serde::Serializer::new(writer))
             .map_err(|e| e.to_string())
     }
+
+    /// Compute the digest of the [VerifierIndex], which can be used for the Fiat-Shamir
+    /// transformation while proving / verifying.
+    pub fn digest<EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>>(
+        &self,
+    ) -> G::BaseField {
+        let mut fq_sponge = EFqSponge::new(self.fq_sponge_params.clone());
+        for comm in self.sigma_comm.iter() {
+            fq_sponge.absorb_g(&comm.unshifted)
+        }
+        for comm in self.coefficients_comm.iter() {
+            fq_sponge.absorb_g(&comm.unshifted)
+        }
+        fq_sponge.absorb_g(&self.generic_comm.unshifted);
+        fq_sponge.absorb_g(&self.psm_comm.unshifted);
+        fq_sponge.absorb_g(&self.complete_add_comm.unshifted);
+        fq_sponge.absorb_g(&self.mul_comm.unshifted);
+        fq_sponge.absorb_g(&self.emul_comm.unshifted);
+        fq_sponge.absorb_g(&self.endomul_scalar_comm.unshifted);
+        fq_sponge.digest_fq()
```

Comment:
I think this function should be called absorb, and you're missing range_check, lookup, chacha, etc. see https://github.com/o1-labs/proof-systems/pull/609/files#diff-c89c45dc84ecd1f409b94d69ca374ed88678982f04f419b79435fd5ebff85b4fR264

Reply:
and you don't need to return the digest IIUC

Reply:
I think I understand what you're doing, you're adding more rows in the verifier circuit but optimizing the absorption of the verifier index by caching its digest. IMO we can have both world by absorbing the verifier index first, and caching the state of the sponge at this point.

Reply:
> and you don't need to return the digest if you pass the sponge and absorb it from here, currently this code recreates a sponge on top of the already existing fq sponge :o

As I mentioned (I think on a call), we want to limit the scope in which we have to use the 'optional sponge' in the verifier circuit. We can do this by using a fresh optional sponge here, squeezing it, and then absorbing the commitment into the regular, non-optional sponge.

> I think I understand what you're doing, you're adding more rows in the verifier circuit but optimizing the absorption of the verifier index by caching its digest. IMO we can have both world by absorbing the verifier index first, and caching the state of the sponge at this point.

We could also cache the state of the main sponge, yes. Looks like it's marginally more efficient to do it this way (we only need to store a single field element rather than the sponge state), but I'm fine with doing either.

Reply:
> you're missing range_check, lookup, chacha, etc

Added these (and expanded the verifier index explicitly to make sure we don't miss any).

Reply:
> the 'optional sponge'

I guess I don't understand what the optional sponge is to understand why this makes sense. Would you mind writing an explanation for it somewhere?